### PR TITLE
Preform physical transforms just before rendering

### LINKED ISSUE #5108: Reduce physical optimization passes; perform just before rendering.
We have many optimizations we perform to `RelationExpr`, though they roughly fall in to the "logical" and "physical" camps. Logical optimizations are oblivious to the physical implementation / rendering process, whereas the physical optimizations bind things like arrangements to use and join strategies. The physical optimizations currently transform the `RelationExpr` in ways that are hard to walk back, for example stating that a certain arrangement should be constructed. We perform physical optimization multiple times, but ideally should perform it only just before rendering, as part of what is currently `optimize_dataflow`.

Related to #5029

### PR DESCRIPTION
**Pull request by asenac**
Add two constructors for the optimizer, with different sets
of transformations: one meant to be used for logical optimization
on views; and one that performs a full optimization pass, including
physical optimizations, meant to be applied on dataflows right
before rendering them.

Fixes #5108

### CODE REVIEW

File: test/sqllogictest/transform/lifting.slt [hunk 1]
```diff
@@ -637,40 +637,51 @@ EOF
...
-| Join %0 %1 (= #0 #3)
-| | implementation = Differential %1 %0.(#0)
+%4 =
+| Union %2 %3
+
+%5 = Let l1 =
+| Join %1 %4 (= #0 #2)
+| | implementation = Differential %4 %1.(#0)
 | | demand = (#0)
 | Filter (#0 <= 195), (#0 >= 38)
```
Comment (wangandi) [inline test/sqllogictest/transform/lifting.slt:667]: These filters are originally around `t_pk`, and the information from them cannot get propagated to `t` until they get lifted after `JoinImplementation` runs. 

#3044 should allow this information to be propagated to `t` before the JoinImplementation stage.

**Review by wangandi — APPROVED** [review]
LGTM. Test regressions look fixable by PredicateKnowledge.

File: test/sqllogictest/transform/lifting.slt (hunk 1 above)
Comment (frankmcsherry) [reply test/sqllogictest/transform/lifting.slt:667]: Sweet; I was unclear on what was going on here; thank you!

**Review by frankmcsherry — APPROVED** [review]
This looks good, thank you! I had imagined more fall out / churn in the query plans, but I guess not! Thanks very much.

File: test/sqllogictest/transform/lifting.slt (hunk 1 above)
Comment (asenac) [reply test/sqllogictest/transform/lifting.slt:667]: Oh, I wasn't aware of #3044. That seems to have some overlap with the new `PredicatePullup` transform in #7471.

File: src/coord/src/coord.rs [hunk 2]
```diff
@@ -3400,7 +3400,7 @@ pub async fn serve(
             let mut coord = Coordinator {
                 worker_guards,
                 worker_txs,
-                optimizer: Default::default(),
+                optimizer: Optimizer::for_view(),
```
[the remark this answers is not in the document]
Comment (asenac) [reply src/coord/src/coord.rs:3403]: `view_optimizer` sounds good to me.

File: src/transform/src/dataflow.rs [hunk 3]
```diff
@@ -15,16 +15,19 @@
...
 
 use dataflow_types::{DataflowDesc, LinearOperator};
-use expr::{Id, LocalId, MirRelationExpr};
+use expr::{GlobalId, Id, LocalId, MirRelationExpr, MirScalarExpr};
 use std::collections::{HashMap, HashSet};
 
 /// Optimizes the implementation of each dataflow.
 ///
 /// This method is currently limited in scope to propagating filtering and
 /// projection information, though it could certainly generalize beyond.
```
[the remark this answers is not in the document]
Comment (asenac) [reply src/transform/src/dataflow.rs]: Oh, right. I missed it. New commit includes an update for the docs.

**Comment by philip-stoev** [comment]
Item No1. This query has an extra `Filter (1234 < #1)` in the prior commit at position # 3 that is missing from this branch:

```
CREATE TABLE IF NOT EXISTS t1 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
EXPLAIN  SELECT  FROM (
	SELECT MAX ( a2 .f2  ) f1  , COUNT (  *  ) f2
	FROM t1 a1
	LEFT JOIN t1 a2 ON TRUE
	WHERE 1234  < a1 .f2
) WHERE EXISTS (
	SELECT WHERE 1 > (SELECT f2)
); ```

the diff in the plans is:

```
--- 6875.out	2021-07-22 10:44:14.005039991 +0200
+++ 16875.out	2021-07-22 10:44:14.065041032 +0200
@@ -15,6 +15,7 @@
                                                  +
  %3 =                                            +
  | Get %2 (l0)                                   +
+ | Filter (1234 < #1)                            + <- this line is missing in this branch
  | Distinct group=(#0, #1)                       +
  | Negate                                        +
```

**Comment by philip-stoev** [comment]
Item No 2. This query is **missing** a `Project (#0, #0, #2)` in this branch:

```
EXPLAIN  SELECT * FROM (SELECT MAX (f2), MAX (f2)  FROM t1) a1  LEFT  JOIN (  SELECT  TRUE  )  ON  TRUE

--- 6875.out	2021-07-22 10:49:24.498367989 +0200
+++ 16875.out	2021-07-22 10:49:24.542368737 +0200
@@ -39,6 +39,7 @@
  %8 =                              +
  | Union %6 %7                     +
  | Map null                        +
+ | Project (#0, #0, #2)            +
                                    +
  %9 =                              +
```

**Comment by philip-stoev** [comment]
Item No 3. Here is a positive change. In this branch, the Filter was pushed before the join:

```
# 2021-07-22T10:32:35 --- /tmp//randgen2752959-1626942755-server0.dump	2021-07-22 10:32:35.484290049 +0200
# 2021-07-22T10:32:35 +++ /tmp//randgen2752959-1626942755-server1.dump	2021-07-22 10:32:35.484290049 +0200
# 2021-07-22T10:32:35 @@ -1,6 +1,5 @@
# 2021-07-22T10:32:35  %0 =
# 2021-07-22T10:32:35  | Get materialize.public.t1 (<U>)
# 2021-07-22T10:32:35 -| Filter (5 >= ((<#> + <#>) + <#>))
# 2021-07-22T10:32:35  | ArrangeBy ()
# 2021-07-22T10:32:35  
# 2021-07-22T10:32:35  %1 =
# 2021-07-22T10:32:35 @@ -11,6 +10,7 @@
# 2021-07-22T10:32:35  | Join %0 %1
# 2021-07-22T10:32:35  | | implementation = Differential %1 %0.()
# 2021-07-22T10:32:35  | | demand = (<#>, <#>)
# 2021-07-22T10:32:35 +| Filter (5 >= ((<#> + <#>) + <#>))
# 2021-07-22T10:32:35  
# 2021-07-22T10:32:35  %3 =
# 2021-07-22T10:32:35  | Get %2 (l0)
```

**Comment by philip-stoev** [comment]
Item No4. This somewhat degenerate query is missing an `ArrangeBy` operator and thus a different join algorithm is used:

```
EXPLAIN     SELECT *   FROM t1  , (  SELECT MIN (  TRUE  )  ,  AVG ( f1  ) AS agg  FROM t1  ) a2 WHERE a2.agg  NOT  IN ( a2 .agg   )
```
```
--- 6875.out	2021-07-22 11:04:16.901883018 +0200
+++ 16875.out	2021-07-22 11:04:16.941883747 +0200
@@ -2,6 +2,7 @@
 ----------------------------------------------------------
  %0 =                                                    +
  | Get materialize.public.t1 (####)                        +
+ | ArrangeBy ()                                          + <- missing in this branch
                                                          +
  %1 =                                                    +
  | Get materialize.public.t1 (####)                        +
@@ -14,7 +15,9 @@
                                                          +
  %2 =                                                    +
  | Join %0 %1                                            +
- | | implementation = Differential %0 %1.()              +
+ | | implementation = DeltaQuery                         +
+ | |   delta %0 %1.()                                    +
+ | |   delta %1 %0.()                                    +
  | | demand = (#0, #1, #4)                               +
  | Map true                                              +
  | Project (#0, #1, #5, #4)                              +
```

**Comment by asenac** [comment]
Item No 1 is a positive change since the filter is redundant as it is already applied in %0 before the join that feeds %3:

```
                                                        Optimized Plan                                                        
------------------------------------------------------------------------------------------------------------------------------
 %0 =                                                                                                                        +
 | Get materialize.public.t1 (u1)                                                                                            +
 | | types = (double precision?, double precision)                                                                           +
 | | keys = ()                                                                                                               +
 | Filter (1234 < #1)                                                                                                        +
 | | types = (double precision?, double precision)                                                                           +
 | | keys = ()                                                                                                               +
 | ArrangeBy ()                                                                                                              +
 | | types = (double precision?, double precision)                                                                           +
 | | keys = ()                                                                                                               +
                                                                                                                             +
 %1 =                                                                                                                        +
 | Get materialize.public.t1 (u1)                                                                                            +
 | | types = (double precision?, double precision)                                                                           +
 | | keys = ()                                                                                                               +
                                                                                                                             +
 %2 = Let l0 =                                                                                                               +
 | Join %0 %1                                                                                                                +
 | | implementation = Differential %1 %0.()                                                                                  +
 | | demand = (#0, #1)                                                                                                       +
 | | types = (double precision?, double precision, double precision?, double precision)                                      +
 | | keys = ()                                                                                                               +
                                                                                                                             +
 %3 =                                                                                                                        +
 | Get %2 (l0)                                                                                                               +
 | | types = (double precision?, double precision, double precision?, double precision)                                      +
 | | keys = ()                                                                                                               +
 | Distinct group=(#0, #1)                                                                                                   +
 | | types = (double precision?, double precision)                                                                           +
 | | keys = ((#0, #1))                                                                                                       +
 | Negate                                                                                                                    +
 | | types = (double precision?, double precision)                                                                           +
 | | keys = ()                                                                                                               +
                                                                                                                             +
```

**Comment by philip-stoev** [comment]
Item No5. In this branch, the Demand list is longer because it seems that the fact that the two columns in the derived table are identical is not recognized.

```
EXPLAIN  SELECT a2  FROM t2  JOIN (  SELECT  AVG ( a1 .f2  )  ,  AVG ( a1 .f2  ) f2  FROM t2 a1  ) a2  ON a2 .f2  NOT  IN (  1   )
```
```
--- 6875.out	2021-07-22 11:08:23.134304441 +0200
+++ 16875.out	2021-07-22 11:08:23.182305292 +0200
@@ -15,7 +15,7 @@
  %2 =                                                    +
  | Join %0 %1                                            +
  | | implementation = Differential %0 %1.()              +
- | | demand = (#4, #4)                                   +
+ | | demand = (#4)                                       +
  | Map record_create(#4, #4)                             +
  | Project (#5)                                          +
  
```

**Comment by asenac** [comment]
Item No 2 is innocuous, the projection of %8 is effectively `[#0, #0, #2]`

```
 %6 =                              +
 | Get %4 (l1)                     +
 | Filter (isnull(#0) || (#0 = #0))+
 | Negate                          +
 | Project (#0, #0)                +
                                   +
 %7 =                              +
 | Get %4 (l1)                     +
 | Filter (isnull(#0) || (#0 = #0))+
 | Project (#0, #0)                +
                                   +
 %8 =                              +
 | Union %6 %7                     +
 | Map null                        +
                                   +
 %9 =                              +
 | Union %5 %8                     +
```

**Comment by philip-stoev** [comment]
Item No6 .This query has a constant plan in the prior commit but is an outer join in this branch:

```
EXPLAIN    SELECT  *  FROM (  SELECT MIN (  5  ) f2  FROM t1  ) a1  WHERE  a1 .f2  =  ANY (   SELECT a1 .f2 FROM (  SELECT  WHERE  FALSE  )  )   OR a1  IS  NULL 
```

**Comment by asenac** [comment]
Item No 4 looks a positive change to me. It used to use a delta query join for a cross join at the cost of creating an arrangement for `t1()`.

**Comment by asenac** [comment]
Item No 5 looks also innocuous to me. `demand` is effectively used as a set. I could make `Demand`, `ProjectionLifting` and a few other transforms use `dedup` to remove duplicates.

**Comment by philip-stoev** [comment]
Item No 7. The location of the ArrangeBy operator has shifted to be earlier in the graph in this branch:

```
  EXPLAIN  SELECT  FROM t2  LEFT  JOIN t2 a2  ON  TRUE  WHERE a2 .f2  IN (   SELECT a2 .f2 c1  FROM (  SELECT  TRUE  )  JOIN t2  ON f1  =  ALL (   SELECT f2  FROM t1  WHERE  3  > (    SELECT f1 agg2  WHERE FALSE     LIMIT  1     )  )  )   

--- 6875.out	2021-07-22 11:32:15.146975518 +0200
+++ 16875.out	2021-07-22 11:32:15.206976538 +0200
@@ -75,7 +75,6 @@
  %16 =                                                +
  | Get %11 (l2)                                       +
  | Filter !(isnull(#0))                               +
- | ArrangeBy (#1)                                     + <- in this branch
                                                       +
  %17 =                                                +
  | Get %15 (l4)                                       +
@@ -92,10 +91,11 @@
                                                       +
  %20 =                                                +
  | Union %17 %19                                      +
+ | ArrangeBy (#0)                                     + <- in the prior commit
                                                       +
  %21 =                                                +
  | Join %16 %20 (= #1 #3)                             +
- | | implementation = Differential %20 %16.(#1)       +
+ | | implementation = Differential %16 %20.(#0)       +
  | | demand = (#0)                                    +
  | Distinct group=(#0)                                +
  | ArrangeBy (#0)                                     +
```

**Comment by asenac** [comment]
Item No 6 is yet another case of `PredicatePushdown` working against `RedundantJoin` and is fixed by #7471. In main. the ArrangeBy operators added during the first pass prevent the filter from being pushed down in the second pass. Since we no longer do join planning in the first passes there are no ArrangeBy operators anymore and the filter does get pushed in the second pass, preventing `RedundantJoin` from doing its job.

**Comment by philip-stoev** [comment]
Item No 8. In this query, the Filter in this branch is after the join whereas in the prior commit it was before the join:

```
EXPLAIN SELECT * FROM lineitem
JOIN orders  ON TRUE
JOIN customer  ON o_custkey  = c_custkey
WHERE c_comment  BETWEEN 'g' AND 'o'
AND c_acctbal  = l_extendedprice;" | ~/diff.sh 
--- 6875.out	2021-07-22 12:05:18.369787593 +0200
+++ 16875.out	2021-07-22 12:05:18.421788445 +0200
@@ -10,12 +10,12 @@
                                                      +
  %2 =                                                +
  | Get materialize.public.customer (####)            +
+ | Filter (#7 <= "o"), (#7 >= "g")                   +
                                                      +
  %3 =                                                +
  | Join %0 %1 %2 (= #5 #30) (= #17 #25)              +
  | | implementation = Differential %2 %0.(#5) %1.(#1)+
  | | demand = (#0..#24, #26..#29, #31, #32)          +
- | Filter (#32 <= "o"), (#32 >= "g")                 +
  | Project (#0..#24, #17, #26..#29, #5, #31, #32)    +
```

The plan depends on the available additional indexes -- with various combinations of additional indexes, both this branch and the prior commit can put the Filter in either position, but with the specific set of indexes below this branch is doing worse than the prior commit:

```
DROP SCHEMA public CASCADE;
CREATE SCHEMA public;

CREATE TABLE lineitem (
    l_orderkey       integer NOT NULL,
    l_partkey        integer NOT NULL,
    l_suppkey        integer NOT NULL,
    l_linenumber     integer NOT NULL,
    l_quantity       decimal(15, 2) NOT NULL,
    l_extendedprice  decimal(15, 2) NOT NULL,
    l_discount       decimal(15, 2) NOT NULL,
    l_tax            decimal(15, 2) NOT NULL,
    l_returnflag     text NOT NULL,
    l_linestatus     text NOT NULL,
    l_shipdate       date NOT NULL,
    l_commitdate     date NOT NULL,
    l_receiptdate    date NOT NULL,
    l_shipinstruct   text NOT NULL,
    l_shipmode       text NOT NULL,
    l_comment        text NOT NULL
);

CREATE TABLE orders (
    o_orderkey       integer,
    o_custkey        integer NOT NULL,
    o_orderstatus    text NOT NULL,
    o_totalprice     decimal(15, 2) NOT NULL,
    o_orderdate      DATE NOT NULL,
    o_orderpriority  text NOT NULL,
    o_clerk          text NOT NULL,
    o_shippriority   integer NOT NULL,
    o_comment        text NOT NULL
);

CREATE INDEX pk_orders_orderkey ON orders (o_orderkey);
CREATE INDEX fk_orders_custkey ON orders (o_custkey ASC);

CREATE TABLE customer (
    c_custkey     integer,
    c_name        text NOT NULL,
    c_address     text NOT NULL,
    c_nationkey   integer NOT NULL,
    c_phone       text NOT NULL,
    c_acctbal     decimal(15, 2) NOT NULL,
    c_mktsegment  text NOT NULL,
    c_comment     text NOT NULL
);

CREATE INDEX pk_customer_custkey ON customer (c_custkey);
```

**Comment by asenac** [comment]
Item No 8 is harmless. It is caused by the change in the order in which `PredicatePushdown` and `JoinImplementation` happen. In this branch the former is happening first, and then `JoinImplemention` lifts the predicate on top of the join. Later, the rendering code does collapse all the filters on top of the join and evaluates them as early as possible IINM.

One extra invocation of `PredicaePushdown` after `JoinImplementation` fixes this _issue_, although it doesn't have any peformance implications:
```
materialize=> EXPLAIN SELECT * FROM lineitem
JOIN orders  ON TRUE
JOIN customer  ON o_custkey  = c_custkey
WHERE c_comment  BETWEEN 'g' AND 'o'
AND c_acctbal  = l_extendedprice;
                    Optimized Plan                    
------------------------------------------------------
 %0 =                                                +
 | Get materialize.public.lineitem (u7)              +
 | ArrangeBy (#5)                                    +
                                                     +
 %1 =                                                +
 | Get materialize.public.orders (u9)                +
 | ArrangeBy (#1)                                    +
                                                     +
 %2 =                                                +
 | Get materialize.public.customer (u13)             +
 | Filter (#7 <= "o"), (#7 >= "g")                   +
                                                     +
 %3 =                                                +
 | Join %0 %1 %2 (= #5 #30) (= #17 #25)              +
 | | implementation = Differential %2 %0.(#5) %1.(#1)+
 | | demand = (#0..#24, #26..#29, #31, #32)          +
 | Project (#0..#24, #17, #26..#29, #5, #31, #32)    +
 
(1 row)

materialize=> 
```

**Review by philip-stoev — APPROVED** [review]
No wrong results, no panics. The plan differences that I examined all fall into the items already reported.

**Comment by wangandi** [comment]
> In this branch the former is happening first, and then `JoinImplementation` lifts the predicate on top of the join.

Why are the filters being lifted given that they belong to %2, and the join implementation does not involve creating an index on %2? 

Is the join implementation changing between loops?

**Comment by asenac** [comment]
> > In this branch the former is happening first, and then `JoinImplementation` lifts the predicate on top of the join.
> 
> Why are the filters being lifted given that they belong to %2, and the join implementation does not involve creating an index on %2?
> 
> Is the join implementation changing between loops?

Good point. I was assuming `JoinImplementation` always lifted all predicates in the join inputs, but as you have implied, it only lifts them when it needs to add an arrangement.

You are right, the  join implementation is changing between loops as follows:

```
%0 =
| Get ? (u1)

%1 =
| Get ? (u3)

%2 =
| Get ? (u7)
| Filter !(isnull(#0)), (#7 <= "o"), (#7 >= "g")

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Unimplemented
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
JoinImplementation (lifts the predicate because it needs to add an arrangement)
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| Filter !(isnull(#25)), (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
ColumnKnowledge
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| Filter true, (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
FoldConstants
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| Filter (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
Demand
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)
| ArrangeBy (#0)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %1.(#1) %2.(#0) %0.(#5)
| | demand = (#0..#24, #26..#29, #31, #32)
| Filter (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)

------------------------------------------------------------------------------------------------------------------------------------------------
JoinImplementation
------------------------------------------------------------------------------------------------------------------------------------------------
%0 =
| Get ? (u1)
| ArrangeBy (#5)

%1 =
| Get ? (u3)
| ArrangeBy (#1)

%2 =
| Get ? (u7)

%3 =
| Join %0 %1 %2 (= #5 #30) (= #17 #25)
| | implementation = Differential %2 %0.(#5) %1.(#1)
| | demand = (#0..#24, #26..#29, #31, #32)
| Filter (#32 <= "o"), (#32 >= "g")
| Project (#0..#24, #17, #26..#29, #5, #31, #32)
```

**Comment by wangandi** [comment]
Indeed it is changing between loops. I'm in favor of merging this PR and creating a followup issue to investigate the necessity multiple invocations of `JoinImplementation`.

**Pull request merged**